source/common/filter/tcp_proxy.cc

```diff
@@ -44,19 +68,19 @@ TcpProxyConfig::TcpProxyConfig(const envoy::api::v2::filter::network::TcpProxy&
   if (config.has_deprecated_v1()) {
     for (const envoy::api::v2::filter::network::TcpProxy::DeprecatedV1::TCPRoute& route_desc :
          config.deprecated_v1().routes()) {
-      if (!context.clusterManager().get(route_desc.cluster())) {
-        throw EnvoyException(
-            fmt::format("tcp proxy: unknown cluster '{}' in TCP route", route_desc.cluster()));
-      }
-      routes_.emplace_back(Route(route_desc));
+      // if (!context.clusterManager().get(route_desc.cluster())) {
+      //   throw EnvoyException(
+      //       fmt::format("tcp proxy: unknown cluster '{}' in TCP route", route_desc.cluster()));
+      // }
+      routes_.emplace_back(Route(route_desc, context.random()));
     }
   }
 
-  if (!config.cluster().empty()) {
-    envoy::api::v2::filter::network::TcpProxy::DeprecatedV1::TCPRoute default_route;
-    default_route.set_cluster(config.cluster());
-    routes_.emplace_back(default_route);
-  }
+  // if (!config.cluster().empty()) {
+  //   envoy::api::v2::filter::network::TcpProxy::DeprecatedV1::TCPRoute default_route;
+  //   default_route.set_cluster(config.cluster());
+  //   routes_.emplace_back(default_route);
+  // }
```

Comment:
@ggreenway any reason for introducing this functionality? According to the proto, tcp_proxy.cluster is just a simple cluster selection (based on the filter chain match criteria). The code above introduces a default fall back cluster functionality which could be done in a more explicit way, instead of overloading the tcp_proxy.cluster.

Reply:
I did this because the proto allows specifying both tcp_proxy.cluster and deprecatedv1 routes.  This was my best guess at how the two should work together.  When the deprecatedv1 routes are removed, this would become much simpler.  

However, to move forward with this PR, would we need to un-deprecate the routes in this filter?

Reply:
Nevermind, I see how this would all fit together with match criteria on listeners instead of deprecatedv1 routes here.

I wonder if this functionality should be attached to the cluster instead of put into tcp_proxy.  I could see other connection types wanting to do weighted routing like this.  Maybe this could be a new cluster type that routes to other clusters by weight or something?

Reply:
Not sure about a cluster. This could certainly be abstracted across http and tcp (much like the access log stuff). But that is a non functional change, that can be done separately. This would let you add timeouts, retries, access logs to all things using the abstract route entry class.